Shiftless Volunteers query perf + integration test

brandonc opened the pull request:
## Overview
Addresses #130 -- Previously, the shiftless volunteers route was loading all volunteers
into memory and performing n<sup>2</sup>+1 database queries to figure
out which ones in the shared regions have no shifts assigned. This was
translated into equivalent SQL

## Details
I thought a count subquery was simpler than an equivalent left join / group / having

## Notes
There remains the question as to whether `current_volunteer.region_ids`
is a decent criteria for finding shiftless volunteers. Another option
might be `current_volunteer.admin_region_ids` ?

rylanb commented on app/controllers/volunteers_controller.rb:
```diff
@@ -32,9 +32,9 @@ def assign
-    end
+    @volunteers = Volunteer.joins(:regions)
+                           .where('regions.id' => current_volunteer.region_ids)
+                           .where('(SELECT COUNT(*) FROM schedule_volunteers WHERE volunteer_id = volunteers.id) = 0')
```
Hmmm ok this looks good based on the code / my limited ability to parse SQL.

My concern is: 
New Version finds: 734 volunteers
Old Version finds: 841 volunteers.

Thoughts?

brandonc replied:
Can you give me a count for this query? I think I should have been counting the schedule chains differently:

```
Volunteer.joins(:regions)
  .where('regions.id' => current_volunteer.region_ids)
  .where('(SELECT COUNT(*) FROM schedule_chains ' \
    'INNER JOIN schedule_volunteers ON ' \
    'schedule_chains.id = schedule_volunteers.schedule_chain_id ' \
    "WHERE schedule_chains.active = 't' " \
    'AND schedule_volunteers.volunteer_id = volunteers.id ' \
    "AND schedule_volunteers.active = 't') = 0")
```
🤢

rylanb replied:
841!

rylanb commented on spec/features/region_admin/shiftless_volunteers_spec.rb:
```diff
@@ -0,0 +1,49 @@
+
+      before do
+        create(:assignment, :admin, volunteer: current_volunteer, region: boulder)
+        create(:assignment, volunteer: shiftless_volunteer, region: boulder)
```
I don't have a full knowledge of the domain and what a shiftless volunteer would be... but wouldn't having an assignment create a shift for the volunteer?

rylanb commented on spec/features/region_admin/shiftless_volunteers_spec.rb:
```diff
@@ -0,0 +1,49 @@
+        create(:assignment, volunteer: shiftless_volunteer, region: boulder)
+        create(:assignment, volunteer: shiftless_volunteer_denver, region: denver)
+        create(:assignment, volunteer: volunteer_with_shifts, region: boulder)
+        create(:schedule_volunteer, volunteer: volunteer_with_shifts, schedule_chain: schedule_chain)
```
Oh maybe its this part that makes 'em scheduled. Strange!

brandonc commented:
Rebased and added a unit test for this query. To answer your question-- an `assignment` is the many-to-many record that relates volunteers to regions.

rylanb commented:
Not confusing at all! I'd say an assignment is a shift! *shakes head*

brandonc commented:
Hang on somehow I may have broken features/shifts_needing_coverage_spec.rb

cllns commented:
It's not very nice to call the volunteers shiftless, just because they don't have an assignment ;)

rylanb commented:
They are adrift in the world, Sean!
